master/internal/api_experiment.go

```diff
@@ -1916,3 +1923,158 @@ func (a *apiServer) GetModelDefFile(
+		Where("new_c.rank = 1")
+
+	checkpointInnerQueryUnion := oldCheckpointOuterQuery.UnionAll(newCheckpointOuterQuery)
+
+	checkpointInnerQuery := db.Bun().NewSelect().
+		TableExpr("(?) as bc", checkpointInnerQueryUnion).
+		ColumnExpr("bc.*").
+		ColumnExpr("row_number() OVER(PARTITION BY bc.trial_id ORDER BY bc.signed_searcher_metric) as order_rank")
+
+	checkpointsQuery := db.Bun().NewSelect().
+		TableExpr("(?) as c", checkpointInnerQuery).
+		Column("c.trial_id").
+		Column("c.resources").
+		ColumnExpr("c.order_rank AS rank").
+		ColumnExpr("c.uuid::text AS uuid").
+		ColumnExpr("c.steps_completed AS total_batches").
+		ColumnExpr("proto_time(c.report_time) AS end_time").
+		ColumnExpr(protoStateDBCaseString(checkpointv1.State_value, "c.state", "state", "STATE_"))
+
+	stepsQuery := db.Bun().NewSelect().
+		TableExpr("steps AS s").
+		Column("s.total_batches").
+		Where("s.trial_id = trials.id").
+		Order("s.total_batches DESC").
+		Limit(1)
+
+	allocationsQuery := db.Bun().NewSelect().
+		TableExpr("allocations AS a").
+		ColumnExpr("extract(EPOCH FROM sum(coalesce(a.end_time, now()) - a.start_time))").
+		Where("a.task_id = trials.task_id")
+
+	trialsInnerQuery := db.Bun().NewSelect().
+		Table("trials").
+		Column("trials.id").
+		Column("trials.experiment_id").
+		Column("trials.runner_state").
+		Column("trials.checkpoint_count").
+		Column("trials.task_id").
+		ColumnExpr("proto_time(trials.start_time) AS start_time").
+		ColumnExpr("proto_time(trials.end_time) AS end_time").
+		ColumnExpr("least(trials.restarts, (ex.config->>'max_restarts')::int) AS restarts").
+		ColumnExpr("coalesce(new_ckpt.uuid, old_ckpt.uuid) AS warm_start_checkpoint_uuid").
+		ColumnExpr("trials.checkpoint_size AS total_checkpoint_size").
+		ColumnExpr(protoStateDBCaseString(trialv1.State_value, "trials.state", "state", "STATE_")).
+		ColumnExpr("(CASE WHEN trials.hparams = 'null'::jsonb THEN null ELSE trials.hparams END) AS hparams").
+		ColumnExpr("(?) AS total_batches_processed", stepsQuery).
+		ColumnExpr("(?) AS wall_clock_time", allocationsQuery).
+		ColumnExpr("row_to_json(lv)::jsonb AS latest_validation").
+		ColumnExpr("row_to_json(bv)::jsonb AS best_validation").
+		ColumnExpr("row_to_json(ch)::jsonb AS best_checkpoint").
+		ColumnExpr("row_number() OVER(PARTITION BY trials.experiment_id ORDER BY (bv.signed_searcher_metric)) as _metric_rank").
+		Join("JOIN ex ON ex.id = trials.experiment_id").
+		Join("LEFT JOIN v bv ON trials.best_validation_id = bv.id").
+		Join("LEFT JOIN v lv ON trials.id = lv.trial_id AND lv.latest_rank = 1").
+		Join("LEFT JOIN raw_checkpoints old_ckpt ON old_ckpt.id = trials.warm_start_checkpoint_id").
+		Join("LEFT JOIN checkpoints_v2 new_ckpt ON new_ckpt.id = trials.warm_start_checkpoint_id").
+		Join("LEFT JOIN ch ON ch.trial_id = trials.id AND ch.rank = 1")
+
+	err = db.Bun().NewSelect().
+		With("ex", experimentValues).
+		With("si", searcherInfoQuery).
+		With("v", validationsQuery).
+		With("ch", checkpointsQuery).
```

Comment:
one thing that has come up in the past is CTEs causing perf problems. see #5744 and #5729

those PRs refer to them as subqueries, but [postgres treats CTEs differently](https://hakibenita.com/be-careful-with-cte-in-postgre-sql#is-it-possible-to-prevent-postgresql-from-materializing-a-cte) than regular subqueries, and i think that is what is actually causing the performance issues

Because this is going to be a frequently-accessed top-level API, that also reaches into some of the bigger tables, we are going to need to have a pretty good understanding of the performance applications, ideally before we start dogfooding the new experience.

I don't we necessarily need to figure it all out for this PR, but I want to make sure we have a path forward that won't require too much reworking later on. It might be that we can replace these `With()` statements with `Join()`s and we are good to go, but I think it's good to at least know what our options are.

Reply:
yeah i guess my concern would just be if it does for all experiments what the trials endpoint does for a single experiment.

Reply:
fair points about the CTEs themselves though